src/python/pants/source/source_root.py

```diff
@@ -191,15 +136,34 @@ def register_options(cls, register):
     def get_pattern_matcher(self) -> SourceRootPatternMatcher:
         return SourceRootPatternMatcher(self.options.root_patterns)
 
-    @memoized_method
-    def get_source_roots(self) -> SourceRoots:
-        # Only v1 code should call this method.
-        return SourceRoots(self.options.root_patterns, self.options.unmatched == "fail")
+
+@dataclass(frozen=True)
+class SourceRootsRequest:
+    """Find the source roots for the given files and/or dirs."""
+
+    files: Tuple[PurePath, ...]
+    dirs: Tuple[PurePath, ...]
```

Comment:
If you expect this type to be used a lot, it would be nice to add a custom constructor that takes `Iterable[PurePath]` then calls `tuple()` on it. See the bottom of https://www.pantsbuild.org/docs/style-guide#prefer-dataclasses for an example.

If this isn't going to be called directly much, this is fine.

Reply:
If you use a custom constructor, you must manually call `self.__post_init__()` as the last line.